### PR TITLE
Harden data and config protection .htaccess

### LINKED ISSUE #6449: Data folder accessible if "Satisfy Any" is set
Rewards are always welcome thou, but it is not me that fell above this: https://help.nextcloud.com/t/htaccess-warning-while-configuration-should-be-ok/20280/17?u=michaing

### Steps to reproduce
1. Set up Nextcloud on Apache2 **without** pretty URLs and data directory **inside** nextcloud root.
2. Ensure `.htaccess` files are used as expected to prevent access to data folder.
3. Add `Satisfy Any` to nextcloud vhost/config file as mentioned in admin manual as necessary in some cases: https://docs.nextcloud.com/server/12/admin_manual/installation/source_installation.html#additional-apache-configurations
4. Try to access to some file inside data folder by using it's direct URL.

### Expected behaviour
Access should be forbidden.

### Actual behaviour
Access works very well.
- Pretty URLs lead to redirection of requests to nextcloud base URL. But e.g. access to `/data/index.html` is still possible.

### Server configuration

**Operating system**: Raspbian/Debian Stretch

**Web server:** Apache/2.4.25

**Database:** MariaDB 10.1

**PHP version:** 7.0.19-1

**Nextcloud version:** 12.0.2

**Updated from an older Nextcloud/ownCloud or fresh install:** updated

**Where did you install Nextcloud from:** downloads.nextcloud.com

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - activity: 2.5.2
  - apporder: 0.4.0
  - calendar: 1.5.3
  - contacts: 1.5.3
  - dav: 1.3.0
  - federatedfilesharing: 1.2.0
  - files: 1.7.2
  - files_sharing: 1.4.0
  - files_trashbin: 1.2.0
  - files_versions: 1.5.0
  - gallery: 17.0.0
  - impersonate: 1.0.1
  - logreader: 2.0.0
  - lookup_server_connector: 1.0.0
  - nextcloud_announcements: 1.1
  - notifications: 2.0.0
  - oauth2: 1.0.5
  - ownnote: 1.08
  - polls: 0.7.3
  - previewgenerator: 1.0.6
  - provisioning_api: 1.2.0
  - serverinfo: 1.2.0
  - sharerenamer: 1.3
  - tasks: 0.9.5
  - twofactor_backupcodes: 1.1.1
  - updatenotification: 1.2.0
  - workflowengine: 1.2.0
Disabled:
  - admin_audit
  - comments
  - encryption
  - federation
  - files_external
  - files_pdfviewer
  - files_texteditor
  - files_videoplayer
  - firstrunwizard
  - imprint
  - password_policy
  - sharebymail
  - survey_client
  - systemtags
  - theming
  - user_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "localhost",
            "my.domain.org"
        ],
        "datadirectory": "\/mnt\/sda\/ncdata", #Tested with manual created data directory + test files inside nextcloud root and with occ maintenance:update:htaccess to create correct .htaccess file inside.
        "dbtype": "mysql",
        "version": "12.0.2.0",
        "memcache.local": "\\OC\\Memcache\\APCu",
        "filelocking.enabled": true,
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "\/var\/run\/redis\/redis.sock",
            "port": 0,
            "dbindex": 0,
            "password": "***REMOVED SENSITIVE VALUE***",
            "timeout": 1.5
        },
        "dbname": "nextcloud",
        "dbhost": "localhost",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "instanceid": "ocv2j0skx6hk",
        "loglevel": 3,
        "logtimezone": "Europe\/Berlin",
        "trashbin_retention_obligation": "disabled",
        "versions_retention_obligation": "disabled",
        "skeletondirectory": "",
        "defaultapp": "apporder",
        "maintenance": false,
        "overwrite.cli.url": "https:\/\/my.domain.org\/nextcloud",
        "htaccess.RewriteBase": "\/nextcloud", #Tested without pretty URLs, as they redirect access tries to all files besides at least index.html inside data directory.
        "mail_smtpmode": "smtp",
        "mail_smtpauthtype": "LOGIN",
        "mail_smtpsecure": "tls",
        "mail_from_address": "my.mail",
        "mail_domain": "gmx.de",
        "mail_smtpauth": 1,
        "mail_smtphost": "mail.gmx.net",
        "mail_smtpport": "587",
        "mail_smtpname": "***REMOVED SENSITIVE VALUE***",
        "mail_smtppassword": "***REMOVED SENSITIVE VALUE***"
    },
    "apps": {
        "activity": {
            "enabled": "yes",
            "installed_version": "2.5.2",
            "notify_email_calendar": "0",
            "notify_email_calendar_event": "0",
            "notify_email_calendar_todo": "0",
            "notify_email_favorite": "0",
            "notify_email_file_changed": "0",
            "notify_email_file_created": "0",
            "notify_email_file_deleted": "0",
            "notify_email_file_downloaded": "0",
            "notify_email_file_restored": "0",
            "notify_email_public_links": "0",
            "notify_email_remote_share": "0",
            "notify_email_shared": "0",
            "notify_setting_batchtime": "604800",
            "notify_setting_self": "1",
            "notify_setting_selfemail": "0",
            "notify_stream_calendar": "1",
            "notify_stream_calendar_event": "1",
            "notify_stream_calendar_todo": "1",
            "notify_stream_favorite": "1",
            "notify_stream_file_changed": "1",
            "notify_stream_file_created": "1",
            "notify_stream_file_deleted": "1",
            "notify_stream_file_downloaded": "1",
            "notify_stream_file_favorite": "0",
            "notify_stream_file_restored": "1",
            "notify_stream_public_links": "1",
            "notify_stream_remote_share": "1",
            "notify_stream_shared": "1",
            "types": "filesystem"
        },
        "apporder": {
            "enabled": "yes",
            "installed_version": "0.4.0",
            "order": "[\"\/nextcloud\/index.php\/apps\/activity\/\",\"\/nextcloud\/index.php\/apps\/files\/\",\"\/nextcloud\/index.php\/apps\/gallery\/\",\"\/nextcloud\/index.php\/apps\/contacts\/\",\"\/nextcloud\/index.php\/apps\/calendar\/\",\"\/nextcloud\/index.php\/apps\/tasks\/\",\"\/nextcloud\/index.php\/apps\/ownnote\/\",\"\/nextcloud\/index.php\/apps\/polls\/\"]",
            "types": ""
        },
        "backgroundjob": {
            "lastjob": "20"
        },
        "calendar": {
            "enabled": "yes",
            "installed_version": "1.5.3",
            "types": ""
        },
        "comments": {
            "enabled": "no",
            "installed_version": "1.2.0",
            "types": "logging"
        },
        "contacts": {
            "enabled": "yes",
            "installed_version": "1.5.3",
            "types": ""
        },
        "core": {
            "backgroundjobs_mode": "cron",
            "installedat": "1496402497.1163",
            "lastcron": "1505128503",
            "lastupdateResult": "[]",
            "lastupdatedat": "1505127887",
            "moveavatarsdone": "yes",
            "oc.integritycheck.checker": "[]",
            "previewsCleanedUp": "1",
            "public_files": "files_sharing\/public.php",
            "public_webdav": "dav\/appinfo\/v1\/publicwebdav.php",
            "scss.variables": "d41d8cd98f00b204e9800998ecf8427e",
            "shareapi_allow_resharing": "no",
            "shareapi_enforce_links_password": "yes",
            "updater.secret.created": "1503506277",
            "vendor": "nextcloud"
        },
        "dav": {
            "enabled": "yes",
            "installed_version": "1.3.0",
            "types": "filesystem"
        },
        "federatedfilesharing": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": ""
        },
        "federation": {
            "enabled": "no",
            "installed_version": "1.2.0",
            "types": "authentication"
        },
        "files": {
            "cronjob_scan_files": "500",
            "enabled": "yes",
            "installed_version": "1.7.2",
            "types": "filesystem"
        },
        "files_downloadactivity": {
            "enabled": "no",
            "installed_version": "1.1.1",
            "types": "filesystem"
        },
        "files_pdfviewer": {
            "enabled": "no",
            "installed_version": "1.1.1",
            "ocsid": "166049",
            "types": ""
        },
        "files_sharing": {
            "enabled": "yes",
            "installed_version": "1.4.0",
            "lookupServerUploadEnabled": "no",
            "types": "filesystem"
        },
        "files_texteditor": {
            "enabled": "no",
            "installed_version": "2.4.1",
            "ocsid": "166051",
            "types": ""
        },
        "files_trashbin": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": "filesystem"
        },
        "files_versions": {
            "enabled": "yes",
            "installed_version": "1.5.0",
            "types": "filesystem"
        },
        "files_videoplayer": {
            "enabled": "no",
            "installed_version": "1.1.0",
            "types": ""
        },
        "firstrunwizard": {
            "enabled": "no",
            "installed_version": "2.1",
            "types": "logging"
        },
        "gallery": {
            "enabled": "yes",
            "installed_version": "17.0.0",
            "types": ""
        },
        "impersonate": {
            "enabled": "yes",
            "installed_version": "1.0.1",
            "types": ""
        },
        "imprint": {
            "content": "test test",
            "enabled": "no",
            "installed_version": "0.2.5",
            "position-guest": "header-right",
            "position-login": "header-right",
            "position-user": "header-right",
            "types": ""
        },
        "logreader": {
            "enabled": "yes",
            "installed_version": "2.0.0",
            "levels": "00011",
            "ocsid": "170871",
            "types": ""
        },
        "lookup_server_connector": {
            "enabled": "yes",
            "installed_version": "1.0.0",
            "types": "authentication"
        },
        "nextcloud_announcements": {
            "enabled": "yes",
            "installed_version": "1.1",
            "pub_date": "Sat, 10 Dec 2016 00:00:00 +0100",
            "types": "logging"
        },
        "notifications": {
            "enabled": "yes",
            "installed_version": "2.0.0",
            "types": "logging"
        },
        "oauth2": {
            "enabled": "yes",
            "installed_version": "1.0.5",
            "types": "authentication"
        },
        "ownbackup": {
            "enabled": "no",
            "installed_version": "17.5.0",
            "types": ""
        },
        "ownnote": {
            "enabled": "yes",
            "folder": "ownNotes",
            "installed_version": "1.08",
            "types": ""
        },
        "password_policy": {
            "enabled": "no",
            "installed_version": "1.2.2",
            "types": ""
        },
        "polls": {
            "enabled": "yes",
            "installed_version": "0.7.3",
            "types": ""
        },
        "previewgenerator": {
            "enabled": "yes",
            "installed_version": "1.0.6",
            "types": "filesystem"
        },
        "provisioning_api": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": "prevent_group_restriction"
        },
        "rainloop": {
            "enabled": "no",
            "installed_version": "5.0.1",
            "rainloop-autologin": "1",
            "types": ""
        },
        "serverinfo": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": ""
        },
        "sharebymail": {
            "enabled": "no",
            "installed_version": "1.2.0",
            "types": "filesystem"
        },
        "sharerenamer": {
            "enabled": "yes",
            "installed_version": "1.3",
            "types": ""
        },
        "survey_client": {
            "enabled": "no",
            "installed_version": "1.0.0",
            "types": ""
        },
        "systemtags": {
            "enabled": "no",
            "installed_version": "1.2.0",
            "types": "logging"
        },
        "tasks": {
            "enabled": "yes",
            "installed_version": "0.9.5",
            "types": ""
        },
        "theming": {
            "enabled": "no",
            "installed_version": "1.3.0",
            "types": "logging"
        },
        "twofactor_backupcodes": {
            "enabled": "yes",
            "installed_version": "1.1.1",
            "types": ""
        },
        "updatenotification": {
            "core": "12.0.2.0",
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": "",
            "update_check_errors": "0"
        },
        "workflowengine": {
            "enabled": "yes",
            "installed_version": "1.2.0",
            "types": "filesystem"
        }
    }
}
```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Opera 49 + Edge 40.15 were tested.

**Operating system:**

#### Web server error log
none

#### Nextcloud log (data/nextcloud.log)
none

#### Browser log
nene

### PR DESCRIPTION
**Pull request by MichaIng**
+ Set "`Satisfy All`" whenever available, as well on Apache 2.4+. This is required to override possible "`Satisfy Any`" on parent dir, which otherwise would allow public access, regardless of "[Require](https://httpd.apache.org/docs/2.4/mod/mod_authz_core.html#require)" directive.
+ Set "`Deny from all`" as well whenever available, to block access regardless of which access control directive takes priority.
+ Assume Apache 2.2 only, if [mod_authz_core](https://httpd.apache.org/docs/2.4/mod/mod_authz_core.html#require) and [mod_access_compat](https://httpd.apache.org/docs/2.4/mod/mod_access_compat.html) are both not available, to avoid doubled directives. In this case set "`Deny from all`" directive only if the providing [mod_authz_host](https://httpd.apache.org/docs/2.2/mod/mod_authz_host.html#deny) module is available. "Satisfy" is a core directive on Apache 2.2.
+ Update Apache version strings. Regarding the used directives/modules, Apache 2.4 and 2.5 (will be released as 2.6) behave the same.
+ Add ordering spaces to better reflect the nested directives and to match style of other `.htaccess` files.
+ <s>Remove one doubled whitespace</s> + some minor syntax "fixes" to match standards:
  - `ifModule` => `IfModule`
  - <s>`$content.= "string";` => `$content .= "string";`</s> **EDIT: Meanwhile done by: https://github.com/nextcloud/server/commit/cdf8c16942c71d0faf71b346beaad70ced22a2fa**

Fixes: https://github.com/nextcloud/server/issues/6449

Signed-off-by: MichaIng <micha@dietpi.com>

### CODE REVIEW

File: lib/private/Setup.php [hunk 1]
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
 		//Require all denied
 		$now =  date('Y-m-d H:i:s');
```
Comment (kesselb) [inline lib/private/Setup.php]:
```suggestion
		$now = date('Y-m-d H:i:s');
```

File: lib/private/Setup.php [hunk 2]
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
...
 		$content = "# Generated by Nextcloud on $now\n";
-		$content.= "# line below if for Apache 2.4\n";
+		$content.= "# Section for Apache 2.4 and 2.5\n";
```
Comment (kesselb) [inline lib/private/Setup.php]:
```suggestion
		$content.= "# Section for Apache 2.4 and 2.6\n";
```

http://httpd.apache.org/dev/devnotes.html

File: lib/private/Setup.php [hunk 3]
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
...
 		$content.= "<ifModule mod_authz_core.c>\n";
-		$content.= "Require all denied\n";
+		$content.= "  Require all denied\n";
+		$content.= "</ifModule>\n";
```
Comment (kesselb) [inline lib/private/Setup.php]:
```suggestion
		$content .= "</ifModule>\n";
```

There should be a whitespace between $content and `.=`. Mind to change it? I know it was like that before.

File: lib/private/Setup.php [hunk 4]
```diff
@@ -542,19 +542,27 @@ public static function protectDataDirectory() {
...
+		$content.= "<ifModule mod_access_compat.c>\n";
+		$content.= "  Deny from all\n";
+		$content.= "  Satisfy All\n";
 		$content.= "</ifModule>\n\n";
-		$content.= "# line below if for Apache 2.2\n";
+		$content.= "# Section for Apache 2.2\n";
 		$content.= "<ifModule !mod_authz_core.c>\n";
```
Comment (kesselb) [inline lib/private/Setup.php]:
```suggestion
		$content.= "<IfModule !mod_authz_core.c>\n";
```

https://httpd.apache.org/docs/2.4/mod/core.html#ifmodule

File: lib/private/Setup.php (hunk 2 above)
Comment (MichaIng) [reply lib/private/Setup.php]: @kesselb 
So do I get it right that 2.5 will be finally released as 2.6?
Should be changed in the autoindex part below as well then.

File: lib/private/Setup.php (hunk 3 above)
Comment (MichaIng) [reply lib/private/Setup.php]: @kesselb 
Makes sense, the whitespace is also used throughout other code as well as `ifModule` => `IfModule` you suggested below. Lets keep it consistent and stay with syntax according to official docs, regarding cases.

File: lib/private/Setup.php (hunk 2 above)
Comment (kesselb) [reply lib/private/Setup.php]: I think so

**Comment by MichaIng** [comment]
Actually it makes sense to add `Order allow,deny` in front of `Deny from all`. This assures that no `Allow` directive for this dir, declared anywhere else, is evaluated after our `Deny` rule. Otherwise theoretically some faulty or harmful config snipped can override our rule and open access.

**Comment by MichaIng** [comment]
@nextcloud/security 
Hey guys, I hope you find some time to look into this and probably we can still merge it with NC17.

I added the `Order Allow,Deny` directive as well. This simply makes sense to assure the `Deny` has always priority. This is as well common across other web applications and their Apache config recommendations and `.htaccess` files to block certain dir access, e.g. [phpBB](https://github.com/phpbb/phpbb/blob/master/phpBB/.htaccess).

**Review by nickvergessen — APPROVED** [review]

**Comment by nickvergessen** [comment]
Well 17 is already packaged. But I guess we can do it for 17.0.1

**Comment by nickvergessen** [comment]
@rullzer @kesselb time to review so we can merge and get this in?

**Comment by MichaIng** [comment]
Resolved conflict caused format fix, which was also included in this PR: https://github.com/nextcloud/server/commit/cdf8c16942c71d0faf71b346beaad70ced22a2fa

**Review by gary-kim — APPROVED** [review]
LGTM

**Pull request merged**